src/wiki/decorators.py

```diff
@@ -71,68 +106,36 @@ def get_article(  # noqa: max-complexity=23
     """
 
     def wrapper(request, *args, **kwargs):
-        from . import models
-
         path = kwargs.pop("path", None)
         article_id = kwargs.pop("article_id", None)
-
-        # fetch by urlpath.path
-        if path is not None:
+        try:
+            article, urlpath = which_article(path, article_id)
+        except NoRootURL:
+            return redirect("wiki:root_create")
+        except models.Article.DoesNotExist:
+            raise Http404("Article id {:} not found".format(article_id))
+        except models.URLPath.DoesNotExist:
             try:
-                urlpath = models.URLPath.get_by_path(path, select_related=True)
-            except NoRootURL:
-                return redirect("wiki:root_create")
-            except models.URLPath.DoesNotExist:
-                try:
-                    pathlist = list(
-                        filter(
-                            lambda x: x != "",
-                            path.split("/"),
-                        )
+                pathlist = list(
+                    filter(
+                        lambda x: x != "",
+                        path.split("/"),
                     )
-                    path = "/".join(pathlist[:-1])
-                    parent = models.URLPath.get_by_path(path)
-                    return HttpResponseRedirect(
-                        reverse("wiki:create", kwargs={"path": parent.path})
-                        + "?slug=%s" % pathlist[-1].lower()
-                    )
-                except models.URLPath.DoesNotExist:
-                    return HttpResponseNotFound(
-                        render_to_string(
-                            "wiki/error.html",
-                            context={"error_type": "ancestors_missing"},
-                            request=request,
-                        )
+                )
+                path = "/".join(pathlist[:-1])
+                parent = models.URLPath.get_by_path(path)
+                return HttpResponseRedirect(
+                    reverse("wiki:create", kwargs={"path": parent.path})
+                    + "?slug=%s" % pathlist[-1].lower()
+                )
+            except models.URLPath.DoesNotExist:
+                return HttpResponseNotFound(
+                    render_to_string(
```

Comment:
The Codecov message above might be relevant.. is it not reaching because of a missing test or because something is wrong in the implementation? :thinking:

Reply:
This is the case when the URLPath exists, but the corresponding article is gone, and its parent URLPath does not exist.

Given that each URLPath points to its parents, this sounds like it's an an impossible case (at least in the absence of race conditions, and writing a test to explicitly check whether whether this race condition is handled gracefully sounds a bit out of scope) – it's ‘handled’ in my patch because I took a very close transformation of the existing decorator, and it seemed to handle this case.

Reply:
Thank you so much for clarifying, I was not able to dwell for long enough to reach this level on enlightenment. But what you say makes absolute sense!